### PR TITLE
Add support for BF2 in connectX mode

### PR DESCRIPTION
**Pull request by SchSeba**
Fix the issue introduced in https://github.com/k8snetworkplumbingwg/sriov-network-operator/pull/240/files#r808624002

Allow to use BF in connectX mode also for OCP platform.
This is needed until we support the systemd configuration

Signed-off-by: Sebastian Sch <sebassch@gmail.com>

### CODE REVIEW

**Comment by SchSeba** [comment]
/cc @adrianchiris @e0ne @zshi-redhat @bn222 @zshi-redhat  @zeeke 

please have a look

File: pkg/plugins/generic/generic_plugin.go [hunk 1]
```diff
@@ -87,12 +87,31 @@ func (p *GenericPlugin) Apply() error {
 			return nil
 		}
 	}
+
+	// Create a map with all the PFs we will need to configure
+	// we need to create it here before we access the host file system using the chroot function
+	// because the SkipConfigVf needs the mstconfig package that exist only inside the sriov-config-daemon file system
+	vfsToConfig := map[string]bool{}
```
Comment (zeeke) [inline pkg/plugins/generic/generic_plugin.go:94]: please, rename this variable (and parameter) to `vfsToSkip` or something like that, as its entries seem to be true when the Virtual Functions must not be configured.

Another solution may be to make `utils.SyncNodeState(...)` receive `ifSpecs []sriovnetworkv1.Interface, ifStatuses []sriovnetworkv1.InterfaceExt` and filter out the to-skip interfcaes before calling it.

File: pkg/utils/utils_mlx.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+	internalCPUIbVportoStatus, exist := mstCurrentData[internalCPUIbVporto]
+	if !exist {
+		return 0, fmt.Errorf("failed to find %s in the mstconfig output command", internalCPUIbVporto)
+	}
+	internalCPUOffloadEngineStatus, exist := mstCurrentData[internalCPUOffloadEngine]
+	if !exist {
+		return 0, fmt.Errorf("failed to find %s in the mstconfig output command", internalCPUOffloadEngine)
+	}
+
+	// check for DPU
```
Comment (wizhaoredhat) [inline pkg/utils/utils_mlx.go:86]: Just a comment: These registers aren't available with some versions of mstflint.

INTERNAL_CPU_MODEL should also be 1 as well. This is true for both DPU mode and NIC mode. Maybe we should add a check here just to be safe.

Comment (SchSeba) [reply pkg/utils/utils_mlx.go:86]: @wizhaoredhat if for both it will be 1 why do we need to check it?

Comment (wizhaoredhat) [reply pkg/utils/utils_mlx.go:86]: It is possible that the NIC was incorrectly set to Separated Host Mode which will have a value of zero.

File: pkg/utils/utils.go [hunk 3]
```diff
@@ -185,17 +190,29 @@ func SyncNodeState(newState *sriovnetworkv1.SriovNetworkNodeState) error {
...
+		mode, err := mellanoxBlueFieldMode(ifStatus.PciAddress)
+		if err != nil {
+			return false, fmt.Errorf("failed to read Mellanox Bluefield card mode for %s,%v", ifStatus.PciAddress, err)
+		}
+
+		if mode == bluefield2Conntexd {
+			return false, nil
+		}
+
+		glog.V(2).Infof("SkipConfigVf(): skip config VF for BF2 device on DPU mode")
```
Comment (e0ne) [inline pkg/utils/utils.go:211]: Neet to confirm that such use-case is not needed

File: pkg/plugins/generic/generic_plugin.go [hunk 4]
```diff
@@ -87,12 +87,31 @@ func (p *GenericPlugin) Apply() error {
 			return nil
 		}
 	}
+
+	// Create a map with all the PFs we will need to configure
+	// we need to create it here before we access the host file system using the chroot function
+	// because the SkipConfigVf needs the mstconfig package that exist only inside the sriov-config-daemon file system
+	vfsToSkip := map[string]bool{}
```
Comment (e0ne) [inline pkg/plugins/generic/generic_plugin.go:94]: nit: it would be good to move this code into the separate function

File: pkg/plugins/generic/generic_plugin.go [hunk 5]
```diff
@@ -87,12 +87,31 @@ func (p *GenericPlugin) Apply() error {
 			return nil
 		}
 	}
+
+	// Create a map with all the PFs we will need to configure
+	// we need to create it here before we access the host file system using the chroot function
```
Comment (e0ne) [inline pkg/plugins/generic/generic_plugin.go:92]: It's clear to my why do we need this logic but it doesn't look good architecture to have this map and still use SkipConfigVf function in other places.

**Review by e0ne — COMMENTED** [review]
I still need to test this PR.  With this change we introduce more platform-specific and NIC-specific code which is used across the operator instead of in s specific plugins

File: pkg/utils/utils_mlx.go (hunk 2 above)
Comment (SchSeba) [reply pkg/utils/utils_mlx.go:98]: sounds good done

File: pkg/plugins/generic/generic_plugin.go (hunk 4 above)
Comment (SchSeba) [reply pkg/plugins/generic/generic_plugin.go]: done

File: pkg/plugins/generic/generic_plugin.go (hunk 5 above)
Comment (SchSeba) [reply pkg/plugins/generic/generic_plugin.go:92]: done we use the same function on both places

**Comment by SchSeba** [comment]
> I still need to test this PR. With this change we introduce more platform-specific and NIC-specific code which is used across the operator instead of in s specific plugins

I agree but I was not able to find a different solution, but I am optimistic that the rework I am doing will make stuff simpler

File: pkg/utils/utils_mlx.go [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+		return bluefield2Dpu, nil
+	} else if strings.Contains(internalCPUPageSupplierstatus, extHostPf) &&
+		strings.Contains(internalCPUEswitchManagerStatus, extHostPf) &&
+		strings.Contains(internalCPUIbVportoStatus, extHostPf) &&
+		strings.Contains(internalCPUOffloadEngineStatus, disabled) &&
+		strings.Contains(internalCPUModelStatus, embeddedCPU) {
+		glog.V(2).Infof("MellanoxBlueFieldMode():card %s in ConnectX mode", PciAddress)
+		return bluefield2Conntexd, nil
+	}
+
```
Comment (wizhaoredhat) [inline pkg/utils/utils_mlx.go:114]: Just a suggestion. I believe it would help with debugging if who have some logs here (in the case of failure) what the actual mstconfig is.

File: pkg/utils/utils_mlx.go [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/golang/glog"
+)
+
+//BlueField2 mode representation
```
Comment (adrianchiris) [inline pkg/utils/utils_mlx.go:11]: the mode and the parameters will be valid for future Bluefield NICs. so maybe drop the "2"

File: pkg/utils/utils_mlx.go [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+type BlueField2Mode int
+
+const (
+	bluefield2Dpu BlueField2Mode = iota
+	bluefield2Conntexd
```
Comment (adrianchiris) [inline pkg/utils/utils_mlx.go:16]: nit "Connexd" -> "ConnectXMode" ?

File: pkg/utils/switchdev.go [hunk 9]
```diff
@@ -29,15 +29,23 @@ func IsSwitchdevModeSpec(spec sriovnetworkv1.SriovNetworkNodeStateSpec) bool {
 }
 
 func WriteSwitchdevConfFile(newState *sriovnetworkv1.SriovNetworkNodeState) (update bool, err error) {
+	// Create a map with all the PFs we will need to SKIP for systemd configuration
+	vfsToSkip, err := GetVfsToSkip(newState)
```
Comment (adrianchiris) [inline pkg/utils/switchdev.go:33]: PFsToSkip/pfsToSkip ?

File: pkg/utils/utils.go (hunk 3 above)
Comment (adrianchiris) [reply pkg/utils/utils.go:211]: what is the use case you refer ?

File: pkg/utils/utils.go [hunk 10]
```diff
@@ -185,17 +190,48 @@ func SyncNodeState(newState *sriovnetworkv1.SriovNetworkNodeState) error {
...
+		return true, nil
 	}
-	return false
+
+	return false, nil
+}
+
+func GetVfsToSkip(ns *sriovnetworkv1.SriovNetworkNodeState) (map[string]bool, error) {
```
Comment (adrianchiris) [inline pkg/utils/utils.go:218]: nit: add docstring

**Review by adrianchiris — COMMENTED** [review]
still need to give this one a deeper look.

one thing in mind is creating an interface to provide NIC information (e,g if its DPU or non DPU)
then keeping the vendor specific logic in plugin.

File: pkg/plugins/mellanox/mellanox_plugin.go [hunk 11]
```diff
@@ -266,25 +258,6 @@ func mlnxNicFromMap(mstData map[string]string) (*mlnxNic, error) {
...
-				regexResult := formatRegex.FindStringSubmatch(line)
-				fwCurrent[attr] = regexResult[3]
-				fwNext[attr] = regexResult[4]
-				break
-			}
-		}
-	}
-	return
-}
-
```
Comment (bn222) [inline pkg/plugins/mellanox/mellanox_plugin.go:287]: This is a refactor, please do that in a separate commit

File: pkg/utils/utils_mlx.go [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+		strings.Contains(internalCPUOffloadEngineStatus, enabled) &&
+		strings.Contains(internalCPUModelStatus, embeddedCPU) {
+		glog.V(2).Infof("MellanoxBlueFieldMode():card %s in DPU mode", PciAddress)
+		return bluefield2Dpu, nil
+	} else if strings.Contains(internalCPUPageSupplierstatus, extHostPf) &&
+		strings.Contains(internalCPUEswitchManagerStatus, extHostPf) &&
+		strings.Contains(internalCPUIbVportoStatus, extHostPf) &&
+		strings.Contains(internalCPUOffloadEngineStatus, disabled) &&
+		strings.Contains(internalCPUModelStatus, embeddedCPU) {
+		glog.V(2).Infof("MellanoxBlueFieldMode():card %s in ConnectX mode", PciAddress)
```
Comment (bn222) [inline pkg/utils/utils_mlx.go:111]: I prefer to have the log happen outside this function

**Comment by SchSeba** [comment]
@adrianchiris about your comment 
> what is the use case you refer ?

what do you mean?

File: pkg/utils/utils_mlx.go (hunk 6 above)
Comment (SchSeba) [reply pkg/utils/utils_mlx.go:114]: done

**Comment by SchSeba** [comment]
@bn222 done

File: pkg/utils/utils_mlx.go (hunk 12 above)
Comment (SchSeba) [reply pkg/utils/utils_mlx.go:111]: I will not like to add another if outside the function. why do you prefer it outside?

File: pkg/plugins/mellanox/mellanox_plugin.go (hunk 11 above)
Comment (SchSeba) [reply pkg/plugins/mellanox/mellanox_plugin.go:287]: done I separate the commit

File: pkg/utils/utils.go [hunk 13]
```diff
@@ -184,18 +189,51 @@ func SyncNodeState(newState *sriovnetworkv1.SriovNetworkNodeState) error {
...
+					glog.Errorf("GetPfsToSkip(): fail to check for skip VFs %s: %v.", iface.PciAddress, err)
+					return vfsToSkip, err
+				}
+				vfsToSkip[iface.PciAddress] = skip
+				break
+			}
+		}
+	}
+
+	return vfsToSkip, nil
```
Comment (bn222) [inline pkg/utils/utils.go:236]: Your function is called `GetPfsToSkip` and you're variable is called `vfsToSkip`

Comment (SchSeba) [reply pkg/utils/utils.go]: done thanks!

**Comment by SchSeba** [comment]
@e0ne @bn222 can you make another round of review so we can merge this PR please :)

File: pkg/utils/utils_mlx.go (hunk 12 above)
Comment (bn222) [reply pkg/utils/utils_mlx.go:111]: Just a preference. I expect that a function that checks the mode doesn't print to the log. The function that is asking to check for the mode could of course check. Separation of concerns imo.

**Review by e0ne — APPROVED** [review]
Verified with vanilla k8s 1.25

**Comment by bn222** [comment]
lgtm

**Comment by SchSeba** [comment]
Thanks for the review folks I am going to merge this PR
